server/src/internalClusterTest/java/org/elasticsearch/snapshots/CorruptedBlobStoreRepositoryIT.java

```diff
@@ -760,28 +763,81 @@ public void testSnapshotWithMissingShardLevelIndexFile() throws Exception {
+                "test-repo",
+                "test-snap-1"
+            ).setWaitForCompletion(true).get();
+            assertEquals(0, restoreSnapshotResponse.getRestoreInfo().failedShards());
+            ensureGreen("test-idx-1", "test-idx-2");
+        }
+
+        logger.info("--> creating another snapshot, which should re-create the missing file");
+        try (
+            var ignored = new BlobStoreIndexShardSnapshotsIntegritySuppressor();
+            var mockLog = MockLog.capture(BlobStoreRepository.class)
+        ) {
+            mockLog.addExpectation(
+                new MockLog.SeenEventExpectation(
+                    "fallback message",
+                    "org.elasticsearch.repositories.blobstore.BlobStoreRepository",
+                    Level.ERROR,
+                    "shard generation [*] in [test-repo][*] not found - falling back to reading all shard snapshots"
+                )
+            );
+            mockLog.addExpectation(
+                new MockLog.SeenEventExpectation(
+                    "shard blobs list",
+                    "org.elasticsearch.repositories.blobstore.BlobStoreRepository",
+                    Level.ERROR,
+                    "read shard snapshots [*] due to missing shard generation [*] in [test-repo][*]"
+                )
+            );
+            CreateSnapshotResponse createSnapshotResponse = clusterAdmin().prepareCreateSnapshot(
+                TEST_REQUEST_TIMEOUT,
+                "test-repo",
+                "test-snap-2"
+            ).setWaitForCompletion(true).setIndices("test-idx-1").get();
+            assertEquals(
+                createSnapshotResponse.getSnapshotInfo().totalShards(),
+                createSnapshotResponse.getSnapshotInfo().successfulShards()
+            );
+            mockLog.assertAllExpectationsMatched();
+        }
+
+        if (randomBoolean()) {
+            indicesAdmin().prepareDelete("test-idx-1").get();
+            RestoreSnapshotResponse restoreSnapshotResponse2 = clusterAdmin().prepareRestoreSnapshot(
+                TEST_REQUEST_TIMEOUT,
+                "test-repo",
+                randomFrom("test-snap-1", "test-snap-2")
+            ).setIndices("test-idx-1").setWaitForCompletion(true).get();
+            assertEquals(0, restoreSnapshotResponse2.getRestoreInfo().failedShards());
+            ensureGreen("test-idx-1", "test-idx-2");
+        }
+
+        logger.info("--> creating another snapshot, which should succeed since the shard gen file now exists again");
         CreateSnapshotResponse createSnapshotResponse = clusterAdmin().prepareCreateSnapshot(
             TEST_REQUEST_TIMEOUT,
             "test-repo",
-            "test-snap-2"
+            "test-snap-3"
         ).setWaitForCompletion(true).setIndices("test-idx-1").get();
-        assertEquals(
-            createSnapshotResponse.getSnapshotInfo().successfulShards(),
-            createSnapshotResponse.getSnapshotInfo().totalShards() - 1
```

Comment:
I liked that we checked that a snapshot would fail, it ensures the file deletion above deleted a relevant file. Can we snapshot `test-idx-2` at the end to verify this still?

Reply:
Not really, at least not without catching an `AssertionError` or suppressing the same assertion we suppressed above. The log messages should be enough here, right?

Reply:
Yeah, sure, the log assertions is enough, thanks.